### PR TITLE
build es6 modules to support docuflow

### PR DESCRIPTION
**Pull request by kevinbarabash**
I'm currently in the process of adding support for rendering examples to [docuflow](https://github.com/khan/docuflow).  Having ES6 modules that can be loaded in the browser will make it easy to compile the examples and run them in the browser.  The ES6 modules produced by this PR are NOT loadable by the browser, but docuflow can easily change that by renaming imports from:
```
import * as React from "react"
```
to
```
import * as React from "/node_modules/react.js"
```
`docuflow` will also rename files and move them in to the appropriate place to be served by the static site.

This configuration is much simpler than what styleguidist and storybook do which is to create complicate custom babel plugins to compile code snippets.

To test this I ran: `yarn run build:es6` and saw the following output:
```
$ yarn run build:es6
yarn run v1.7.0
$ rollup -c rollup.config.js

packages/wonder-blocks-button/index.js → packages/wonder-blocks-button/dist/es/index.js...
(!) The 'objectWithoutProperties' Babel helper is used more than once in your code. It's strongly recommended that you use the "external-helpers" plugin or the "es2015-rollup" preset. See https://github.com/rollup/rollup-plugin-babel#configuring-babel for more information
(!) The 'extends' Babel helper is used more than once in your code. It's strongly recommended that you use the "external-helpers" plugin or the "es2015-rollup" preset. See https://github.com/rollup/rollup-plugin-babel#configuring-babel for more information
created packages/wonder-blocks-button/dist/es/index.js in 209ms

packages/wonder-blocks-color/index.js → packages/wonder-blocks-color/dist/es/index.js...
created packages/wonder-blocks-color/dist/es/index.js in 55ms

packages/wonder-blocks-core/index.js → packages/wonder-blocks-core/dist/es/index.js...
created packages/wonder-blocks-core/dist/es/index.js in 220ms

packages/wonder-blocks-dropdown/index.js → packages/wonder-blocks-dropdown/dist/es/index.js...
created packages/wonder-blocks-dropdown/dist/es/index.js in 236ms

packages/wonder-blocks-form/index.js → packages/wonder-blocks-form/dist/es/index.js...
created packages/wonder-blocks-form/dist/es/index.js in 125ms

packages/wonder-blocks-grid/index.js → packages/wonder-blocks-grid/dist/es/index.js...
created packages/wonder-blocks-grid/dist/es/index.js in 59ms

packages/wonder-blocks-icon/index.js → packages/wonder-blocks-icon/dist/es/index.js...
created packages/wonder-blocks-icon/dist/es/index.js in 26ms

packages/wonder-blocks-icon-button/index.js → packages/wonder-blocks-icon-button/dist/es/index.js...
created packages/wonder-blocks-icon-button/dist/es/index.js in 43ms

packages/wonder-blocks-layout/index.js → packages/wonder-blocks-layout/dist/es/index.js...
created packages/wonder-blocks-layout/dist/es/index.js in 11ms

packages/wonder-blocks-link/index.js → packages/wonder-blocks-link/dist/es/index.js...
created packages/wonder-blocks-link/dist/es/index.js in 53ms

packages/wonder-blocks-modal/index.js → packages/wonder-blocks-modal/dist/es/index.js...
created packages/wonder-blocks-modal/dist/es/index.js in 133ms

packages/wonder-blocks-progress-spinner/index.js → packages/wonder-blocks-progress-spinner/dist/es/index.js...
created packages/wonder-blocks-progress-spinner/dist/es/index.js in 14ms

packages/wonder-blocks-spacing/index.js → packages/wonder-blocks-spacing/dist/es/index.js...
created packages/wonder-blocks-spacing/dist/es/index.js in 5ms

packages/wonder-blocks-toolbar/index.js → packages/wonder-blocks-toolbar/dist/es/index.js...
created packages/wonder-blocks-toolbar/dist/es/index.js in 20ms

packages/wonder-blocks-tooltip/index.js → packages/wonder-blocks-tooltip/dist/es/index.js...
created packages/wonder-blocks-tooltip/dist/es/index.js in 148ms

packages/wonder-blocks-typography/index.js → packages/wonder-blocks-typography/dist/es/index.js...
created packages/wonder-blocks-typography/dist/es/index.js in 122ms
✨  Done in 2.06s.
```

One of the things I found super helpful about rollup's output is that it indicated when it couldn't find dependencies.  This allowed me to fix a couple of missing dependencies in some of our packages.

It also flagged a couple of issue with the duplicate code that babel was adding that we'll want to address at some point.  I can only assume that the webpack build is doing the same thing b/c they're both using babel under the hood.

> (!) The 'objectWithoutProperties' Babel helper is used more than once in your code. It's strongly recommended that you use the "external-helpers" plugin or the "es2015-rollup" preset. See https://github.com/rollup/rollup-plugin-babel#configuring-babel for more information
> (!) The 'extends' Babel helper is used more than once in your code. It's strongly recommended that you use the "external-helpers" plugin or the "es2015-rollup" preset. See https://github.com/rollup/rollup-plugin-babel#configuring-babel for more information

Here's a short snippet of the type of code that's generated for our modules, in this case `wonder-blocks-button`:
```
import React__default, { Component, createElement } from 'react';
import { StyleSheet } from 'aphrodite';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { LabelLarge, LabelSmall } from '@khanacademy/wonder-blocks-typography';
import Color, { SemanticColor, mix, fade } from '@khanacademy/wonder-blocks-color';
import { addStyle, getClickableBehavior } from '@khanacademy/wonder-blocks-core';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }


const StyledAnchor = addStyle("a");
const StyledButton = addStyle("button");
const StyledLink = addStyle(Link);

class ButtonCore extends React__default.Component {
    constructor(...args) {
        var _temp;

        return _temp = super(...args), this.handleClick = e => {
            if (this.props.disabled) {
                e.preventDefault();
            }
        }, _temp;
    }

    render() {
        const _props = this.props,

...
```

Notice that the imports have not been renamed.  That's `docuflow`'s job.

### CODE REVIEW

File: packages/wonder-blocks-core/util/ssr-id-factory.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // @flow
-import {IIdentifierFactory} from "../util/types.js";
+import type {IIdentifierFactory} from "../util/types.js";
```
Comment (kevinbarabash) [inline packages/wonder-blocks-core/util/ssr-id-factory.js]: rollup complained about this but webpack didn't which I fond a bit surprising.

File: packages/wonder-blocks-button/package.json [hunk 2]
```diff
@@ -7,6 +7,7 @@
   },
   "description": "",
   "main": "dist/index.js",
+  "module": "dist/es/index.js",
```
Comment (kevinbarabash) [inline packages/wonder-blocks-button/package.json]: Using the `module` field for `ES6` modules is one of the many ways of doing this (there's also `jsnext`).  rollup supports consuming modules with either.  I decided to go with `module` b/c it sounded nicer.

File: packages/wonder-blocks-core/util/ssr-id-factory.js (hunk 1 above)
Comment (jeresig) [reply packages/wonder-blocks-core/util/ssr-id-factory.js]: Huh, fascinating!

File: packages/wonder-blocks-toolbar/package.json [hunk 3]
```diff
@@ -15,7 +16,8 @@
   "license": "MIT",
   "dependencies": {
     "@khanacademy/wonder-blocks-color": "^1.0.6",
-    "@khanacademy/wonder-blocks-core": "^1.1.1"
+    "@khanacademy/wonder-blocks-core": "^1.1.1",
+    "@khanacademy/wonder-blocks-typography": "^1.0.6"
```
Comment (jeresig) [inline packages/wonder-blocks-toolbar/package.json]: Did rollup catch this, too?

File: packages/wonder-blocks-form/package.json [hunk 4]
```diff
@@ -4,6 +4,8 @@
   "design": "v1",
   "description": "Form components for Wonder Blocks.",
   "main": "dist/index.js",
+  "module": "dist/es/index.js",
+  "source": "index.js",
```
Comment (jeresig) [inline packages/wonder-blocks-form/package.json]: Can you add a check to the pre-publish-check.js that confirms that module and source are set to the right values? (In addition to the `main` check that I have right now.)

File: packages/wonder-blocks-button/package.json (hunk 2 above)
Comment (jeresig) [reply packages/wonder-blocks-button/package.json]: Seems good to me!

File: .circleci/config.yml [hunk 5]
```diff
@@ -31,7 +31,8 @@ jobs:
           paths:
             - node_modules
       - run: yarn run gen-snapshot-tests
-      - run: yarn run build:all
+      - run: yarn run build:cjs
+      - run: yarn run build:es6
```
Comment (jeresig) [inline .circleci/config.yml]: Why do we have to build these in circleci? (Honest question, It's not clear to me if they're being used at all.)

File: package.json [hunk 6]
```diff
@@ -8,8 +8,9 @@
     "clean": "rm -rf packages/wonder-blocks-*/dist && rm -rf packages/wonder-blocks-*/node_modules",
     "lint": "eslint --config ./eslint/eslintrc packages",
     "lint:watch": "esw --watch --config ./eslint/eslintrc packages",
-    "build:all": "webpack --display-modules && node utils/package-flow-types.js",
-    "watch:all": "webpack -w",
+    "build:cjs": "webpack --display-modules && node utils/package-flow-types.js",
```
Comment (jeresig) [inline package.json]: You could also keep a `build:all` which then runs both `build:cjs` and `build:es6`.

File: rollup.config.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import fs from "fs";
```
Comment (jeresig) [inline rollup.config.js]: I haven't used rollup before but this seems pretty intuitive to me!

**Pull request merged**